src/seer.erl

```diff
@@ -13,59 +15,63 @@
          read/2,
          read_all/0]).
 
--spec counter_inc(binary()) -> ok.
+-spec counter_inc(metric_name()) -> ok.
 counter_inc(Counter) ->
     counter_inc(Counter, 1).
 
--spec counter_inc(binary(), integer()) -> ok.
+-spec counter_inc(metric_name(), integer()) -> ok.
 counter_inc(Counter, Val) ->
     {ok, Ref} = get_or_new(counter, Counter),
     atomics:add(Ref, 1, Val).
 
--spec dist_record(binary(), integer()) -> ok.
+-spec dist_record(metric_name(), integer()) -> ok.
 dist_record(Dist, Val) ->
     {ok, Ref} = get_or_new(dist, Dist),
     N = atomics:add_get(Ref, ?DIST_RESERVOIR_SIZE + 1, 1),
     dist_record2(Ref, Val, N).
 
--spec dist_timing(binary(), erlang:timestamp()) -> ok.
+-spec dist_timing(metric_name(), erlang:timestamp()) -> ok.
 dist_timing(Timing, TimeStamp) ->
     Delta = timer:now_diff(os:timestamp(), TimeStamp),
     dist_record(Timing, Delta).
 
--spec gauge_set(binary(), integer()) -> ok.
+-spec gauge_set(metric_name(), integer()) -> ok.
 gauge_set(Gauge, Val) ->
     {ok, Ref} = get_or_new(gauge, Gauge),
     atomics:put(Ref, 1, Val).
 
--spec histo_record(binary(), integer()) -> ok.
+-spec histo_record(metric_name(), integer()) -> ok.
 histo_record(Histo, Val) ->
     {ok, Ref} = get_or_new(histo, Histo),
     histo_record2(Ref, Val).
 
--spec histo_timing(binary(), erlang:timestamp()) -> ok.
+-spec histo_timing(metric_name(), erlang:timestamp()) -> ok.
 histo_timing(Histo, TimeStamp) ->
     Delta = timer:now_diff(os:timestamp(), TimeStamp),
     histo_record(Histo, Delta).
 
+-spec read(metric_type(), metric_name()) -> {ok, read_value()}.
 read(Type, Name) ->
     {ok, Ref} = get(Type, Name),
     {ok, read_metric(Type, Name, Ref)}.
 
+-spec read_all() -> [{metric_type(), metric_name(), read_value()}].
 read_all() ->
     [{Type, Name, read_metric(Type, Name, Ref)} || {{?MODULE, Name}, {Type, Ref}} <- persistent_term:get()].
 
 % private
+-spec dist_record2(atomics:atomics_ref(), integer(), integer()) -> ok.
 dist_record2(Ref, Val, N) when N =< ?DIST_RESERVOIR_SIZE ->
     atomics:put(Ref, N, Val);
-dist_record2(Ref, Val, N) ->
+dist_record2(Ref, Val, N) when N >= 1 andalso N =< 4294967295 ->
```

Comment:
As it is right now, attempting to `dist_record` with a value greater than 2^32 will cause a `function_clause` exception and crash. Are we ok with this?

Reply:
I'd say we are not! I see three alternatives: Ignore the values past the maximum,  keep logging new values using the lowest probability or use `rand:uniform(N)` when higher values are reached. However, since the goal of using granderl was to go faster, and since a high N probably means the logging rate is extremely high, using a slower random method seems counter productive, so I'd favour one of the two first options.

Reply:
Another possibility might be to punt the concern to the user by having `dist_record` return `ok | {error, term()}`. If the user cares, they can pattern match on the `ok` result and crash otherwise. If they don't care, they can forgo the pattern matching and throw away the result of the call.

Reply:
`{error, overflow}` is now returned when N is too high. The value is not used.